### PR TITLE
Associations-Project

### PR DESCRIPTION
**Pull request by DanSam5K**
In this Project
- I Set-up .github workflows actions with rubocop and stylelint linters
- Create event model and set up authentication with devise to handle user model
- Create enrollment model and add all the necessary views HTML.erb for all model and controller action
- Add more to column to models when needed and perform all required database migrations
- Implement more features that will allow users to invite and enrol for events and cancel the invitation
- Modify project readme and fix linters errors

### CODE REVIEW

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :red_circle: :stop_sign: 
---------------
Hello @DanSam5K, great job working on this project :100: :+1: 
TO HIGHLIGHTS: 

1. Your Project is professional 
2. Your repo is professional
3. Well structured files


However, there are a few changes to make to complete this project :hammer_and_wrench: see them below :eyes: :point_down: 
-------------------

- [x] When I try to attend an event I got an error. 

Screenshot from 2021-06-29 19-53-04

And when I try to invite someone I cannot send it. Please fix this. 

- [x] For the above I cannot see the list of attendees. 
- [x] Please add a user's show page. Displays events the user has created and events that the user is attending to separated in past and upcoming
- [x] In your `events_controller.rb` in `index` action. Lists past events, i.e. `@past_events = Event.previous_events` and Lists upcoming events, i.e. `@upcoming_events = Event.upcoming_events`

past_upcoming_events_requirement (1)

- [x] In your `user_controller.rb` in `show` action 
 

- Lists events created by user, i.e. `@created_events = current_user.events` 

current_user_show_requirement (1)

-  Lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous`

-  Lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` 

user_events_requirement (2)

- [x] Please provide the full command in the steps to clone the repo on your Readme file. 
- [x] Please add test `rspec` and integrations tests `capybara` 

integration_tests_requirement

unit_tests_requirement
----------------------
Make the required changes and send it again for another review :muscle: :100: 
HAPPY CODING!!!

File: README.md [hunk 1]
```diff
@@ -1,24 +1,65 @@
...
+- Clone this project with the command `git clone https://github.com/DanSam5K/Private-Events.git`
+- `cd Private-Events` folder
+- Run `bundle install` in your terminal
+- Run `rails db:migrate` in your terminal
+- Run `rails s` in your terminal to start your local host.
+- In your terminal go to (http://127.0.0.1:3000/) 
+- Sign up for a session and create-events "secrets" or you can just watch them anonymously!
 
-* ...
+Now, you can start interratcing with application
```
Comment (Georjane) [inline README.md]: - [x] I think you have a typo here. It should be **_interacting_**. I advise you to use Grammarly to avoid such typos

**Review by Georjane — CHANGES_REQUESTED** [review]
### Hello @DanSam5K, ✋🏽 

Great work on making some of the changes requested by the previous reviewer 👏 but there are still some which aren't addressed yet.

## Required Changes ♻️

- [x] Signout is not working. It throws an error,
image

- [x] Upcoming event is different in two pages, `user/show` and `/events.1`

`user/show`
image
`/events.1`
image

- [x] I was not able to see the list of attendees, as the previous code reviewer because I was unable to sign out and create another user. Please can you check this?
- [x]  In your user_controller.rb in show action, 
- lists events created by user, i.e. `@created_events = current_user.events`
- lists events previously attended by user using event model scope, i.e. `@past_events = current_user.attended_events.previous`
- lists upcoming events previously attended by user using event model scope, i.e. `@upcoming_events = current_user.attended_events.upcoming`

- [x] Please add test integrations tests `capybara`
- [x] Please check the comment below!

Cheers and Happy coding! 💻 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 🤝 

_Code Reviewer
**[Jane](https://witahgeorjane.tech/)**_

File: README.md (hunk 1 above)
Comment (DanSam5K) [reply README.md]: Noted Jane! Will give it my best to improve the project.

**Comment by DanSam5K** [comment]
implemented the changes requried

**Review by MkrtichSargsyan — COMMENTED** [review]
## _Status:_ Changes Requested :recycle:

## Hello Dansam Samuel

## To Highlight :tada:

1. Great work making the necessary corrections from the previous reviews.
2. You have descriptive git commit messages.

However, there are some changes that you need to make in order to get your project approved

<hr>

- [x] Event's index page displays past and upcoming events lists [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png)

- [x] I couldn't invite a user to the event. This page is opening after clicking on the invite button

aacs


- [x] A user can not join to event 

- [x] In your user_controller.rb in show action

   1. lists events created by user, i.e. @created_events = current_user.events
   2. lists events previously attended by user using event model scope, i.e. @past_events = current_user.attended_events.previous
   3. lists upcoming events previously attended by user using event model scope, i.e. @upcoming_events = current_user.attended_events.upcoming
   
- [x] Please add more tests,also don't forget about integrations tests ```capybara```

Please, make the REQUIRED changes and submit your request again! :grinning:
Good luck. 

> ps: . If you have any questions about my review or need help, you can message me on slack ( @mkrtichsargsyan )

**Review by GzimAsani — CHANGES_REQUESTED** [review]
# STATUS: Changes Required :red_circle: 

## Hello @DanSam5K I know that you are working hard and you are almost there :+1: 

### To improve: 

- [x] Refer to the previous reviewer about his requests in the `user_controller`  `show` action
- [x] Now, when I log in as a user1 create an event and invite user2 to that event and then log in as user2 and try to enroll in the event I get the error `Your name is not on the invitation list`. Kindly fix this

### If you have any question with regards to  the requirements add a comment here and I will reply ASAP :thought_balloon: 

# HAPPY CODING :computer: :desktop_computer:

**Comment by DanSam5K** [comment]
Hello, Thank you that is because you will need to accept the invitation then enrol.

**Comment by GzimAsani** [comment]
> Hello, Thank you that is because you will need to accept the invitation then enrol.

Aha, that is my mistake. I apologize

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    @user = User.current_user.build(event_params)
+    @user.save
+    session[:user_id] = @user.id
+    flash.notice = "User #{@user.name} successfully created! Thank you for signing up!"
+    redirect_to user_path(@user)
+  end
+
+  def show
+    @all_events = Event.all
+    @events = Event.where(user_id: current_user.id)
```
Comment (cyonii) [inline app/controllers/users_controller.rb]: ### MANDATORY

- [x] Lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] Lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.past` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] Lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

Basically, the user `#show` view should display events that the user has **"created"**, and the events the user is attending, **"separated"** by **"past"** and **"upcoming"**. You can use the query above to get those data

**Review by cyonii — CHANGES_REQUESTED** [review]
### STATUS: CHANGES REQUIRED ♻️✍️

Hello team, your resilience is impressive, keep it on.

#### TO IMPROVE (MANDATORY)
You seem to be missing the same request from 2 reviews ago, I hope I can make it clearer. 
- [x] See inline comment

#### NICE TO HAVE (OPTIONAL)
- [x] You heavily exposed your application logic in your views, the previous reviewers probably missed that. There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers.

Kindly implement the changes requested above and resubmit. Happy coding 🤓💻.

> #### <br>Feel free to leave a comment or question if something is not clear. <br> Slack: @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR)<br><br>

**Review by mke2111 — APPROVED** [review]
# 😄 🥇  STATUS: APPROVED  🥇 👏🏽 

## Hey Happy Coder, 👋🏽 
Congratulations, 🥂 . You got minimum requirements for approval. 👍🏽 

### **_Well Done_** 💯 

    1. Your effort towards this project has been great. Keep up the energy.

### **_Room for Improvement_** 🤪 

    1. Remove all commented-out code from your controllers.
    2. You may choose to fix all optional issues that have been mentioned previously. This will improve your understanding of the core principles and best practices of programming.

### You may now merge to the develop branch and remember to submit a project completion form.
### I wish you the very best in your upcoming projects.

**Cheers 🍹 and Happy coding 💻** 

<hr>

### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

**Comment by DanSam5K** [comment]
God Bless you Roy @mke211.
It means a lot to me have been stuck here for weeks😅😄 😅 .

- Will improve the optional changes as advice.🤝
